src/relay/pass/quantize.cc

```diff
@@ -0,0 +1,546 @@
+        Expr ret = (*f)(n->expr, static_cast<int>(kQInput));
+        return static_cast<Expr>(QAnnotateExprNode::make(ret, kQInput));
+      }
+      return e;
+    };
+  return ForwardRewrite(expr, "FQAnnotateRewrite", nullptr, nullptr);
+});
+
+
+// =============
+// realize pass
+
+Expr QRealizeIntExprNode::Realize() const {
+  const auto& cfg = QConfig::Current();
+  Expr data = this->data;
+  if (cfg->store_lowbit_output) {
+    data = Cast(data, cfg->dtype_input);
+  }
+  // dequantize
+  data = Cast(data, Float(32));
+  data = Multiply(data, this->dom_scale);
+  return data;
+}
+
+QRealizeIntExpr QRealizeIntExprNode::make(Expr data, Expr dom_scale, DataType dtype) {
+  NodePtr<QRealizeIntExprNode> n = make_node<QRealizeIntExprNode>();
+  n->data = std::move(data);
+  n->dom_scale = std::move(dom_scale);
+  n->dtype = std::move(dtype);
+  return QRealizeIntExpr(n);
+}
+
+
+inline Expr ForwardOp(const Call& ref_call, const Array<Expr>& args) {
+  return CallNode::make(ref_call->op,
+    args, ref_call->attrs, ref_call->type_args);
+}
+
+
+/* calculate `data * s1 / s2`, use shift if possible */
+inline Expr MulAndDiv(Expr data, float s1, float s2) {
+  // here we assume the dtype of data is dtype activation
+  const QConfig& cfg = QConfig::Current();
+  if (s1 == s2) return data;
+
+  float factor = s1 / s2;
+  float shift_factor = std::log2(factor);
+  if (static_cast<int>(shift_factor) == shift_factor) {
+    LOG(INFO) << "left shift " << shift_factor;
+    return LeftShift(data, MakeConstantScalar(cfg->dtype_activation, static_cast<int>(shift_factor)));
+  } else if (static_cast<int>(factor) == factor) {
+    return Multiply(data, MakeConstantScalar(cfg->dtype_activation, factor));
+  } else {
+    LOG(FATAL) << "fall back to float computation";
+    data = Cast(data, Float(32));
+    return Multiply(data, MakeConstantScalar(Float(32), factor));
+  }
+}
+
+Expr QuantizeRealize(const Call& ref_call,
+                     const Array<Expr>& new_args,
+                     const NodeRef& ctx) {
+  const QConfig& cfg = QConfig::Current();
+  // do not handle data type cast
+  const auto param = ref_call->attrs.as<SimulatedQuantizeAttrs>();
+  CHECK_EQ(param->rounding, "round");
+
+  Expr dom_scale = new_args[1];
+  Expr clip_min = new_args[2];
+  Expr clip_max = new_args[3];
+
+  float dom_scale_imm = GetScalarFromConstant<float>(dom_scale);
+  float clip_min_imm = GetScalarFromConstant<float>(clip_min);
+  float clip_max_imm = GetScalarFromConstant<float>(clip_max);
+
+  // x * idom_scale = y * odom_scale
+  // => y = x * idom_scale / odom_scale
+  if (const auto* n = new_args[0].as<QRealizeIntExprNode>()) {
+    Expr data = n->data;
+    float idom_scale_imm = GetScalarFromConstant<float>(n->dom_scale);
+    float odom_scale_imm = GetScalarFromConstant<float>(dom_scale);
+    float shift_nbit = std::log2(odom_scale_imm / idom_scale_imm);
+    // int32->int8
+    if (static_cast<int>(shift_nbit) == shift_nbit) {
```

Comment:
You may need to specially handle case when shift_nbit == 0, and shift_nbit < 0, or use CHECK to assert they will not happen